File: problem11/cpp/matrix_product.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <iomanip>

const int N = 20;
const int ADJ = 4;

typedef std::vector<std::vector<int> > Matrix;

// Reads the matrix from a file
bool read_matrix(const std::string& filename, Matrix& matrix) {
    std::ifstream file(filename);
    if (!file) return false;
    matrix.resize(N, std::vector<int>(N));
    for (int i = 0; i < N; ++i)
        for (int j = 0; j < N; ++j)
            if (!(file >> matrix[i][j])) return false;
    return true;
}

// Returns the greatest product of four adjacent numbers in any direction
int greatest_product(const Matrix& matrix, std::vector<std::pair<int, int> >& coords) {
    int max = 0;
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            // right
            if (j + ADJ <= N) {
                int prod = 1;
                for (int k = 0; k < ADJ; ++k) prod *= matrix[i][j+k];
                if (prod > max) {
                    max = prod;
                    coords.clear();
                    for (int k = 0; k < ADJ; ++k) coords.emplace_back(i, j+k);
                }
            }
            // down
            if (i + ADJ <= N) {
                int prod = 1;
                for (int k = 0; k < ADJ; ++k) prod *= matrix[i+k][j];
                if (prod > max) {
                    max = prod;
                    coords.clear();
                    for (int k = 0; k < ADJ; ++k) coords.emplace_back(i+k, j);
                }
            }
            // diag down-right
            if (i + ADJ <= N && j + ADJ <= N) {
                int prod = 1;
                for (int k = 0; k < ADJ; ++k) prod *= matrix[i+k][j+k];
                if (prod > max) {
                    max = prod;
                    coords.clear();
                    for (int k = 0; k < ADJ; ++k) coords.emplace_back(i+k, j+k);
                }
            }
            // diag down-left
            if (i + ADJ <= N && j - ADJ + 1 >= 0) {
                int prod = 1;
                for (int k = 0; k < ADJ; ++k) prod *= matrix[i+k][j-k];
                if (prod > max) {
                    max = prod;
                    coords.clear();
                    for (int k = 0; k < ADJ; ++k) coords.emplace_back(i+k, j-k);
                }
            }
        }
    }
    return max;
}

void print_matrix_with_highlight(const Matrix& matrix, const std::vector<std::pair<int, int> >& coords) {
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            bool highlight = false;
            for (size_t idx = 0; idx < coords.size(); ++idx) {
                const std::pair<int, int>& p = coords[idx];
                if (p.first == i && p.second == j) highlight = true;
            }
            if (highlight)
                std::cout << "\033[31m" << std::setw(2) << std::setfill('0') << matrix[i][j] << "\033[0m ";
            else
                std::cout << std::setw(2) << std::setfill('0') << matrix[i][j] << " ";
        }
        std::cout << "\n";
    }
}

int main(int argc, char* argv[]) {
    std::string filename = argc > 1 ? argv[1] : "../matrix";
    Matrix matrix;
    if (!read_matrix(filename, matrix)) {
        std::cerr << "Failed to read matrix from " << filename << std::endl;
        return 1;
    }
    std::vector<std::pair<int, int> > coords;
    int max = greatest_product(matrix, coords);
    print_matrix_with_highlight(matrix, coords);
    std::cout << "Greatest product of four adjacent numbers: " << max << std::endl;
    return 0;
}

File: problem11/matrix_product.py
def read_matrix(filename):
    with open(filename) as f:
        return [list(map(int, line.strip().split())) for line in f if line.strip()]

def greatest_product_in_matrix(matrix, n):
    max_product = 0
    rows = len(matrix)
    cols = len(matrix[0]) if rows else 0
    for r in range(rows):
        for c in range(cols):
            # Horizontal
            if c + n <= cols:
                product = 1
                for i in range(n):
                    product *= matrix[r][c+i]
                max_product = max(max_product, product)
            # Vertical
            if r + n <= rows:
                product = 1
                for i in range(n):
                    product *= matrix[r+i][c]
                max_product = max(max_product, product)
            # Diagonal
            if r + n <= rows and c + n <= cols:
                product = 1
                for i in range(n):
                    product *= matrix[r+i][c+i]
                max_product = max(max_product, product)
            # Anti-diagonal
            if r + n <= rows and c - n + 1 >= 0:
                product = 1
                for i in range(n):
                    product *= matrix[r+i][c-i]
                max_product = max(max_product, product)
    return max_product

def main():
    import sys
    if len(sys.argv) < 2:
        print("Usage: python matrix_product.py <matrix_file>")
        sys.exit(1)
    matrix = read_matrix(sys.argv[1])
    result = greatest_product_in_matrix(matrix, 4)
    print(f"Greatest product of four adjacent numbers: {result}")

if __name__ == "__main__":
    main()
